[Merge] Implemented the autoscaling steps to be proportions of the max replicas.

Templum opened the pull request:
Implemented the proportional autoscaling steps based on the maxReplicas as described in #569. Additionally, I added support for a label `com.openfaas.scale.factor` which allows configuring the proportion, setting it to 0 even allows to disable autoscaling. 

Signed-off-by: Simon Pelczer <templum.dev@gmail.com>

## Description
I updated the signature for `CalculateReplicas` to also take in `scalingFactor uint64` which will be used to calculate the scaling step. This value can be set using a label `com.openfaas.scale.factor`, which also can be used to disable auto-scaling (by setting it to zero). I updated the existing tests for `alerthandler.go` to reflect the done changes. Additionally, I created a dedicated test for the disable functionality when setting `scalingFactor = 0 `.

## Motivation and Context
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
-  #569 

## How Has This Been Tested?
- Executed all of the tests located within the gateway

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly. [See here](https://github.com/openfaas/docs/pull/5)
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

alexellis commented:
@stefanprodan please could you review this change to the auto-scaler?

Templum commented:
As an minor information and to avoid any confusion I always do my commits with Gitkraken. Therefore It shows two commiters.

alexellis commented:
Awaiting some comments from Stefan.

stefanprodan commented:
@Templum can you publish your gateway image to Docker Hub and post it here please? 
Have you tested this with a load testing tool?

Templum commented:
@stefanprodan  Thanks for looking into it, I have my docker image available under `$docker pull templum/faas-gateway`.  I did not have a load testing tool. However, I reused the [repository ](https://github.com/Templum/OpenFaaS-Gracefully-Shutdown-Test) for this. I changed the faas gateway for mine. And also removed the time window so it runs endlessly. And as expected it scaled in 4 increments `( Scaling Factor / 100 * Max Replicas) (Max Replicas: 20 / Scaling Factor : 20)`

alexellis commented:
> I did not have a load testing tool

Have a look at the `hey` tool. I mention how to use it here - https://blog.alexellis.io/serverless-databases-with-openfaas-and-mongo/

alexellis commented on gateway/plugin/external.go:
```diff
@@ -111,12 +113,28 @@ func (s ExternalServiceQuery) GetReplicas(serviceName string) (handlers.ServiceQ
 			}
 		}
+
+		if len(scaleFactor) > 0 {
```
Let's extract this to a function of its own that can take a default value as an override.

Templum replied:
It has more "logic" attached to it than the other labels, so this definitely makes sense.  I will go ahead and move it into a function.

Templum replied:
I will go ahead and do that change. It makes sense.

alexellis commented on gateway/plugin/external.go:
```diff
@@ -154,3 +146,20 @@ func (s ExternalServiceQuery) SetReplicas(serviceName string, count uint64) erro
+
+// This methods tries to parse the provided raw label value and if it fails
+// it will return the provided fallback value and log and message
+func extractLabelValue(rawLabelValue string, fallback uint64 ) uint64{
```
Looks like a good refactor.

Comments on functions (when given) should start with the name of the function i.e. `extractLabelValue will parse the provided raw label...`

Templum replied:
Okay, that was something I didn't know. I will fix that then.

alexellis commented on gateway/plugin/external.go:
```diff
@@ -85,38 +85,30 @@ func (s ExternalServiceQuery) GetReplicas(serviceName string) (handlers.ServiceQ
-		}
+		minReplicas = extractLabelValue(labels[handlers.MinScaleLabel], minReplicas)
+		maxReplicas = extractLabelValue(labels[handlers.MaxScaleLabel], maxReplicas)
+		temp := extractLabelValue(labels[handlers.ScalingFactorLabel], scalingFactor)
```
How about calling `temp` `scalingFactor`. temp is not a descriptive name?

Templum replied:
Sure `temp` is not descriptive. But using `scalingFactor ` would be not a good idea as the provided value could still be in an illegal range and in that case we would require a fallback to the default value, which we would have overridden. I will make the name more descriptive.

alexellis commented:
Simon, this is looking good so thanks for taking the time. Final point before merging would be to ask you to edit the last commit message. The subject should be a maximum of 50-72 characters with a blank line followed by a message explaining the change. You should be able to do this easily via the CLI and the `git commit -s --amend` command.

If you look at the UI for GitHub you'll see the message has been truncated half-way through a word. Here's an example in our contribution guide - https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md#commit-messages

<img width="816" alt="screen shot 2018-04-03 at 22 07 43" src="https://user-images.githubusercontent.com/6358735/38275962-7a65d93a-378b-11e8-881c-281faff8caa8.png">

alexellis commented:
Once merged let's update the docs.openfaas.com page which talks about auto-scaling. I'm looking forward to trying this out :tada:

alexellis commented:
@Templum I opened this up to do a merge and found a merge error on `gateway/tests/alerthandler_test.go` - sorry but I think you need to rebase again. @kenfdev had a change come in that conflicted.

Templum commented:
@alexellis I did the rebase and yeah honestly every one of my commits is signed-off. Have no clue what the bot means. Do you know which commit it refers to ?!

BurtonR commented:
Hey @Templum it looks like you did a `merge` of master into your branch, which then included _all_ commits from master to yours. What you need to do is `rebase` to stack only you commits at the head of master. I've been there, done that too

Templum commented:
@BurtonR I did the rebase however, the merge was the issue. Now it should be fine, thanks for pointing it out :) 👍

viveksyngh commented:
It is also recommended to squash the multiple commits into a single commit for a pull request.

Templum commented:
But the sqashing can be done by the merger or should I already do it ?
